### DISCUSSION TITLE
run code in the Spike robot

### DISCUSSION THREAD

**Discussion by Xylar52** [pybricks/support#984] [Q&A]
is it possible to, like EV3, choose and run code directly from the spike prime robot ? Now, the only thing you can do is run the most recent file, but on ev3 and spike lego firmware you can choose from up to 15 file. I really need this because i play in the First Lego League and while in the game, you can't use you pc to launch code.

**Comment by dlech** [comment]
Once you download and run the program once using Pybricks Code, you can use the button on the hub to start the program again. When you power off the hub, the program is saved to non-volatile memory so it will still be there where you turn it back on.

![image](https://user-images.githubusercontent.com/963645/224368104-19393040-9fb7-4523-8d3a-b5eb4581e129.png)

**Comment by laurensvalk** [comment]
Yes, you can make your own menu with multiple programs. Here is an example:
https://github.com/pybricks/support/issues/861

You can choose numbers, letters, icons, or anything else to make it easy to remember what each program does.

**Reply by thehomelessguy** [reply]
I am embarrassed to say we didn't know about the "%" operator and our code was much larger as a result. Will have to have the kids edit things. Very nice example that I wish we saw months ago as we ended up essentially doing that on our own.

**Comment by thehomelessguy** [comment]
Not quite ready to share our code yet. We are using pybricks for our FLL team. We actually wrote our code modular enough that it runs on both technic and spike hubs. We made technic hub versions of our modular attachment bot that take the same attachments and are identical except for the hub (obviously) and the fact that the technic hub bots lack the light sensors (they only have 4 ports so we just did the motors). These work great as development bots as the kids can just remote control them to build new attachments/hats (we call our attachment sets hats as, with our robot design, they kind of sit over the top of the robot).

We wrote a class we call RobotProgram that has attributes of function (holds the callable that has the program code in it), button color (color of the hub button), character (single ASCII character for the hub to display), and an attribute that holds an object of the button class(more on this later). 

We then extended the spike hub class built into pybricks to add two new attributes. One is a slot that hold a list of items of the RobotProgram class. The second attribute slot holds a number that indicates which item in that list is active. We then wrote methods to go forwards/backwards through the list as well as methods to properly update the hub/display based on the current active program. These were all added to the default PrimeHub class. The color attribute is kind of fun as, for our FLL competition, we are using 4 different "hats" and have it so that the color of the hat required for the current program (we have a lot of technic beams) matches the color displayed by the hub. Kind of a fun human factors bit. 

Really nice for FLL as you can have the programs auto advance. Our team only ever has to push one button to between runs, no loading things up, having to select the next program, etc. 

In terms of the technic hubs, we are able to call our various programs from the remotes. We added an attribute to the default hub class to also store a list of RobotProgram objects. Our remote also can handle various button mappings. We have one setup that simply controls the robot (again incredibly useful for figuring out the FLL missions/building attachments) but also have a remote mode where when a button is pressed it checks the list of programs and runs the one that matches the button press (hence why our RobotProgram class has a button attribute). We can actually run 4/6 of our programs using the technic hubs that we use in our FLL competition in such a manner (2 of our programs use line following/squaring) and we actually have the kids using the cheap technic hubs for much of our development process. 

90% of the code is the same between the technic and spike hub versions of our robots. Just the Hub classes and Robot classes are different (we also wrote Robot classes that has attributes that hold the various classes that make up the robot/have methods specific to the robot. For example, the PrimeRobot class has the methods for line following, line squaring and line finding. Both bots have an attribute that hold our extended DriveBase class, an AttachBase class we wrote, etc (Having all the compenents be _____Base was something we devised to make things easier for beginning coders).. 

We are debating sharing the code at the end of the season. Our team is Exploding Toaster Supernova #57674. As you can tell by our very large number, we are a first-year team but have loved using pybricks. Currently scoring 400/410 points if everything goes right in about 2:25 so pybricks has been very good to us. Not able to get the truck to its parking spot in time so we are giving up on the last 10 points (but we do have it coded).
